pallas-miniprotocols/src/blockfetch/client.rs

```diff
@@ -0,0 +1,171 @@
+    NoBlocks,
+
+    #[error("error while sending or receiving data through the channel")]
+    ChannelError(ChannelError),
+}
+
+pub type Body = Vec<u8>;
+
+pub struct Client<H>(State, ChannelBuffer<H>)
+where
+    H: Channel,
+    Message: Fragment;
+
+impl<H> Client<H>
+where
+    H: Channel,
+    Message: Fragment,
+{
+    pub fn new(channel: H) -> Self {
+        Self(State::Idle, ChannelBuffer::new(channel))
+    }
+
+    pub fn state(&self) -> &State {
+        &self.0
+    }
+
+    pub fn is_done(&self) -> bool {
+        self.0 == State::Done
+    }
+
+    fn has_agency(&self) -> bool {
+        match self.state() {
+            State::Idle => true,
+            State::Busy => false,
+            State::Streaming => false,
+            State::Done => false,
+        }
+    }
+
+    fn assert_agency_is_ours(&self) -> Result<(), Error> {
+        if !self.has_agency() {
+            Err(Error::AgencyIsTheirs)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn assert_agency_is_theirs(&self) -> Result<(), Error> {
+        if self.has_agency() {
+            Err(Error::AgencyIsOurs)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn assert_outbound_state(&self, msg: &Message) -> Result<(), Error> {
+        match (&self.0, msg) {
+            (State::Idle, Message::RequestRange { .. }) => Ok(()),
+            (State::Idle, Message::ClientDone) => Ok(()),
+            _ => Err(Error::InvalidOutbound),
+        }
+    }
+
+    fn assert_inbound_state(&self, msg: &Message) -> Result<(), Error> {
+        match (&self.0, msg) {
+            (State::Busy, Message::StartBatch) => Ok(()),
+            (State::Busy, Message::NoBlocks) => Ok(()),
+            (State::Streaming, Message::Block { .. }) => Ok(()),
+            (State::Streaming, Message::BatchDone) => Ok(()),
+            _ => Err(Error::InvalidInbound),
+        }
+    }
+
+    pub fn send_message(&mut self, msg: &Message) -> Result<(), Error> {
+        self.assert_agency_is_ours()?;
+        self.assert_outbound_state(msg)?;
+        self.1.send_msg_chunks(msg).map_err(Error::ChannelError)?;
+
+        Ok(())
+    }
+
+    pub fn recv_message(&mut self) -> Result<Message, Error> {
+        self.assert_agency_is_theirs()?;
+        let msg = self.1.recv_full_msg().map_err(Error::ChannelError)?;
+        self.assert_inbound_state(&msg)?;
+
+        Ok(msg)
+    }
+
+    pub fn send_request_range(&mut self, range: (Point, Point)) -> Result<(), Error> {
+        let msg = Message::RequestRange { range };
+        self.send_message(&msg)?;
+        self.0 = State::Busy;
+
+        Ok(())
+    }
+
+    pub fn recv_request_range(&mut self) -> Result<(), Error> {
+        match self.recv_message()? {
+            Message::StartBatch => {
+                self.0 = State::Streaming;
+                Ok(())
+            }
+            Message::NoBlocks => {
+                self.0 = State::Idle;
+                Err(Error::NoBlocks)
```

Comment:
Is it correct to error for something which is technically a valid state transition? Maybe `request_range` can Error if it receives a NoBlocks because it's like a higher-level method, but `recv_request_range` is the only "low-level" function we have to progress the state machine. But perhaps this is the best Rust-y way as I'm not sure what the alternative return type would be. Not a big deal really.

Reply:
I went through the same mental process that you did. At first I created a custom enum to handle both results (started and no-blocks) as an Ok() variant, but it seemed too verbose for something that was easily handled by an extra error variant.

I think that in most scenarios, receiving no blocks after requesting a range would result in an error.

I like the idea of using a more strict approach that represents the protocol without any assumptions, but then hiding this into a higher-level abstraction as a different method.

Reply:
moved the error to `fetch_single` and `fetch_range` wrapper methods.